shared/src/ledger/storage/merkle_tree.rs

```diff
@@ -156,81 +253,98 @@ impl<H: StorageHasher + Default> MerkleTree<H> {
         }
     }
 
-    fn tree(
-        &self,
-        store_type: &StoreType,
-    ) -> &SparseMerkleTree<H, H256, DefaultStore<H256>> {
+    fn tree(&self, store_type: &StoreType) -> Either<&Smt<H>, &Amt<H>> {
         match store_type {
-            StoreType::Base => &self.base,
-            StoreType::Account => &self.account,
-            StoreType::Ibc => &self.ibc,
-            StoreType::PoS => &self.pos,
+            StoreType::Base => Either::Left(&self.base),
+            StoreType::Account => Either::Left(&self.account),
+            StoreType::Ibc => Either::Right(&self.ibc),
+            StoreType::PoS => Either::Left(&self.pos),
         }
     }
 
     fn update_tree(
         &mut self,
         store_type: &StoreType,
-        key: H256,
-        value: H256,
+        key: MerkleKey<H>,
+        value: Either<Hash, TreeBytes>,
     ) -> Result<()> {
-        let tree = match store_type {
-            StoreType::Account => &mut self.account,
-            StoreType::Ibc => &mut self.ibc,
-            StoreType::PoS => &mut self.pos,
+        let sub_root = match store_type {
+            StoreType::Account => self
+                .account
+                .update(key.try_into()?, value.unwrap_left())
+                .map_err(Error::MerkleTree)?,
+            StoreType::Ibc => self
+                .ibc
+                .update(key.try_into()?, value.unwrap_right())
+                .map_err(Error::MerkleTree)?,
+            StoreType::PoS => self
+                .pos
+                .update(key.try_into()?, value.unwrap_left())
+                .map_err(Error::MerkleTree)?,
             // base tree should not be directly updated
             StoreType::Base => unreachable!(),
         };
-        let sub_root = tree.update(key, value).map_err(Error::Smt)?;
 
         // update the base tree with the updated sub root without hashing
         if *store_type != StoreType::Base {
             let base_key = H::hash(&store_type.to_string());
-            self.base.update(base_key, *sub_root)?;
+            self.base.update(base_key.into(), Hash::from(sub_root))?;
         }
         Ok(())
     }
 
     /// Check if the key exists in the tree
     pub fn has_key(&self, key: &Key) -> Result<bool> {
         let (store_type, sub_key) = StoreType::sub_key(key)?;
-        let subtree = self.tree(&store_type);
-        let value = subtree.get(&H::hash(sub_key.to_string()))?;
-        Ok(!value.is_zero())
+        let value = match self.tree(&store_type) {
+            Either::Left(smt) => {
+                smt.get(&H::hash(sub_key.to_string()).into())?.is_zero()
+            }
+            Either::Right(amt) => {
+                let key =
+                    StringKey::try_from_bytes(sub_key.to_string().as_bytes())?;
+                amt.get(&key)?.is_zero()
+            }
+        };
```

Comment:
`Either::map_left` and `Either::map_right`, then `Either::into_inner`, to get a `Result<bool>`